### PR TITLE
Preserve DSv2 metadata writes and reinserts for SPARK-50820 [databricks]

### LINKED ISSUE #15348: [AI-AUDIT][SPARK-50820][SQL] Preserve DSv2 row-level DML metadata and reinsert operations in GPU writes
## Upstream Spark change

- JIRA: SPARK-50820
- Spark commit: `8313320ab0136d1e7e5c6cc7638899eb82e42d75`
- Subject: `[SPARK-50820][SQL] DSv2: Conditional nullification of metadata columns in DML`
- First stable Spark release with the fix: Apache Spark 4.1.0
- Audit source: branch-4.1 at `d52a8e13efde20fb9ae0e36b116516b61e279eaa`

Spark added metadata-aware row-level DML write handling for DSv2 tables. The CPU path now selects `DataAndMetadataWritingSparkTask` for `ReplaceDataExec` when `ReplaceDataProjections.metadataProjection` is present, handles `WRITE_WITH_METADATA_OPERATION`, and routes `REINSERT_OPERATION` to `DeltaWriter.reinsert(...)` with metadata when available. The Spark change notes this is needed for row lineage in Iceberg and Delta, where some metadata columns must be preserved and others nullified during UPDATE/MERGE.

## cudf-spark concern

The GPU Spark 4.1 write path appears to still use the older behavior:

- `sql-plugin/src/main/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala`: `GpuReplaceDataWritingSparkTask` projects only `projs.rowProjection` and calls `writer.write(projected)`. It does not use `metadataProjection` or call `writer.write(metadata, row)`.
- `sql-plugin/src/main/spark350/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala`: `GpuDeltaWritingSparkTask` and `GpuDeltaWithMetadataWritingSparkTask` group `REINSERT_OPERATION` rows through `DeltaInsertFilter.filterInsertRows(batch)` and call `writer.insert(...)`, not `writer.reinsert(...)`.
- `sql-plugin/src/main/spark400/scala/com/nvidia/spark/rapids/shims/DeltaInsertFilter.scala` recognizes `REINSERT_OPERATION`, but the downstream GPU task still treats those rows as inserts.

## Impact

Severity: P0 correctness suspect.

For DSv2 row-level tables that depend on conditional metadata nullification or distinguish insert from reinsert, GPU execution may produce different table metadata/row-lineage state from CPU Spark. That can affect future DML behavior and query results.

## Suggested pytest

Add coverage under `integration_tests/src/main/python/iceberg/` for an Iceberg row-lineage MERGE or UPDATE path that requires preserved/nullified metadata, then compare CPU and GPU table contents plus row-lineage or snapshot metadata.

If current Iceberg test infrastructure does not expose the row-lineage metadata cleanly, add a small test-only DSv2 row-level table under `integration_tests/src/main/scala` whose writer records whether Spark calls `write(metadata, row)` and `reinsert(metadata, row)`. Drive it from pytest and assert GPU matches CPU for `WRITE_WITH_METADATA_OPERATION` and `REINSERT_OPERATION`.

### PR DESCRIPTION
**Pull request by firestarman**
Fixes #15348.

### Description

- Route GPU `ReplaceDataExec` through metadata-aware writing tasks when `ReplaceDataProjections.metadataProjection` is present, so Spark 4.0-4.1 `WRITE_WITH_METADATA_OPERATION` rows and Spark 4.2 `UPDATE_OPERATION`/`COPY_OPERATION` rows call `writer.write(metadata, row)` like the CPU path.
- Preserve input ordering by processing contiguous operation runs independently, because globally regrouping metadata and ordinary writes can revisit partitions that clustered writers have already closed.
- Split Delta delete, update, insert, and reinsert handling into separate retry scopes, so an OOM in a later operation cannot replay writer side effects that already succeeded.
- Route `REINSERT_OPERATION` rows through `DeltaWriter.reinsert(...)` with metadata when available, instead of treating reinserts as ordinary inserts and diverging from SPARK-50820 semantics.
- Keep this change connector-agnostic and omit Iceberg V3 row-lineage integration because Iceberg V3 is not currently supported by the plugin.
- Add `GpuDsv2MetadataWriteSuite` coverage for metadata-aware task selection, ordered contiguous runs, retry isolation after a completed metadata write, and insert/reinsert routing with metadata.
- Validated Spark 4.0 / Scala 2.13 scalastyle with `mvn -s /home/liangcail/.m2/settings_art.xml -f scala2.13/pom.xml -Dbuildver=400 -Dcuda.version=cuda13 -DskipTests validate`.
- Compiled the Spark 4.0 plugin and tests and ran `GpuDsv2MetadataWriteSuite` with `mvn -s /home/liangcail/.m2/settings_art.xml -f scala2.13/pom.xml -Dbuildver=400 -Dcuda.version=cuda13 -pl sql-plugin,tests -am test -DwildcardSuites=org.apache.spark.sql.execution.datasources.v2.GpuDsv2MetadataWriteSuite`; all 5 tests passed in both ANSI modes.
- Compiled the Spark 4.2 plugin with `mvn -s /home/liangcail/.m2/settings_art.xml -f scala2.13/pom.xml -Dbuildver=420 -Dcuda.version=cuda13 -DskipTests -pl sql-plugin -am compile`.

### Checklists

Documentation
- [ ] Updated for new or modified user-facing features or behaviors
- [x] No user-facing change

Testing
- [x] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
      (Please provide the names of the existing tests in the PR description.)
- [ ] Not required

Performance
- [ ] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [x] Not required

### CODE REVIEW

File: sql-plugin/src/main/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala [hunk 1]
```diff
@@ -73,3 +86,54 @@ case class GpuReplaceDataWritingSparkTask(
...
+    withResource(withMetadataFilter) { _ =>
+      withResource(rowProjection.project(batch)) { rows =>
+        val dataBatch = GpuColumnVector.filter(rows, rowDataTypes, withMetadataFilter)
+        closeOnExcept(dataBatch) { _ =>
+          if (dataBatch.numRows() > 0) {
+            val metadataBatch = withResource(metadataProjection.project(batch)) { metadata =>
+              GpuColumnVector.filter(metadata, metadataDataTypes, withMetadataFilter)
+            }
+            GpuDsv2WriteMetadata.withMetadataSchema(metadataProj.schema) {
+              writer.write(metadataBatch, dataBatch)
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala]: Fixed in the latest commit: `GpuDataAndMetadataWritingSparkTask.write` now wraps projection/filter/write in `withRetryNoSplit(batch)`, matching `GpuDeltaWritingSparkTask` / `GpuDeltaWithMetadataWritingSparkTask`.

File: sql-plugin/src/main/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala [hunk 2]
```diff
@@ -73,3 +86,58 @@ case class GpuReplaceDataWritingSparkTask(
...
+  private lazy val rowDataTypes = dataProj.schema.fields.map(_.dataType)
+  private lazy val metadataProjection = GpuProjectingColumnarBatch(metadataProj)
+  private lazy val metadataDataTypes = metadataProj.schema.fields.map(_.dataType)
+
+  override protected def write(
+      writer: DataWriter[ColumnarBatch],
+      batch: ColumnarBatch): Unit = {
+    // Match GpuDelta*WritingSparkTask: projection/filter/write allocate on GPU and must
+    // participate in spill-and-retry. withRetryNoSplit closes `batch` on success.
+    withRetryNoSplit(batch) { _ =>
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala]: Fixed: metadata-row and ordinary-row writes now use separate `withRetryNoSplit` scopes over a shared `SpillableColumnarBatch`, so a retryable OOM in the ordinary-row phase cannot replay a completed `write(metadata, data)`.

**Comment by firestarman** [comment]
build

File: sql-plugin/src/main/spark350/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala [hunk 3]
```diff
@@ -569,16 +563,16 @@ case class GpuDeltaWithMetadataWritingSparkTask(
...
-            val filterRows = GpuColumnVector.filter(rows, rowDataTypes, insertFilter)
-            if (filterRows.numRows() > 0) {
-              writer.insert(filterRows)
-            } else {
-              filterRows.close()
-            }
-          }
+        // INSERT and REINSERT must be routed separately (SPARK-50820):
+        // insert(...) vs reinsert(metadata, row).
+        DeltaInsertFilter.writeInserts(writer, batch, rowProjection, rowDataTypes)
```
Comment (res-life) [inline sql-plugin/src/main/spark350/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: [MUST FIX] `writeInserts` and `writeReinserts` still run inside the same `withRetryNoSplit` callback as the preceding delete/update calls. If filtering or projection for a later operation OOMs, retry replays writer side effects that already succeeded; these mutations are not idempotent. Apply the same fix to the non-metadata task above.

Suggested change:
```scala
withResource(SpillableColumnarBatch(
    batch, SpillPriorities.ACTIVE_ON_DECK_PRIORITY)) { scb =>
  // Extract the current operation bodies into these helpers.
  withRetryNoSplit { withResource(scb.getColumnarBatch())(writeDeletes(writer, _)) }
  withRetryNoSplit { withResource(scb.getColumnarBatch())(writeUpdates(writer, _)) }
  withRetryNoSplit { withResource(scb.getColumnarBatch()) { b =>
    DeltaInsertFilter.writeInserts(writer, b, rowProjection, rowDataTypes)
  }}
  withRetryNoSplit { withResource(scb.getColumnarBatch()) { b =>
    DeltaInsertFilter.writeReinserts(
      writer, b, rowProjection, rowDataTypes, metadataProjection, metadataDataTypes)
  }}
}
```

File: sql-plugin/src/main/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala [hunk 4]
```diff
@@ -73,3 +88,77 @@ case class GpuReplaceDataWritingSparkTask(
...
+  private lazy val metadataProjection = GpuProjectingColumnarBatch(metadataProj)
+  private lazy val metadataDataTypes = metadataProj.schema.fields.map(_.dataType)
+
+  override protected def write(
+      writer: DataWriter[ColumnarBatch],
+      batch: ColumnarBatch): Unit = {
+    // Separate retry scopes per writer side-effect. A single withRetryNoSplit around both
+    // phases would replay a completed write(metadata, data) if ordinary-row filtering OOMs.
+    withResource(SpillableColumnarBatch(batch, SpillPriorities.ACTIVE_ON_DECK_PRIORITY)) { scb =>
+      withRetryNoSplit {
```
Comment (res-life) [inline sql-plugin/src/main/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala]: [MUST FIX] Running all metadata rows before all ordinary rows changes the input operation order. For partition-sorted input, this can write partitions `P0,P1` in the first phase and then revisit `P0` in the second phase, which Iceberg's clustered writer rejects after closing that partition.

Suggested change:
```scala
// Split only where column 0 changes; do not group equal operations globally.
withResource(splitIntoContiguousOperationRuns(batch)) { runs =>
  runs.foreach { run =>
    withRetryNoSplit {
      withResource(run.getColumnarBatch()) { b =>
        run.operation match {
          case WRITE_WITH_METADATA_OPERATION => writeMetadataRows(writer, b)
          case WRITE_OPERATION => writeOrdinaryRows(writer, b)
        }
      }
    }
  }
}
```
Please implement `splitIntoContiguousOperationRuns` with an existing stable contiguous-split primitive and preserve ownership of every produced batch.

File: sql-plugin/src/main/spark420/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    inner: SparkPlan,
+    refreshCache: () => Unit,
+    projections: ReplaceDataProjections,
+    write: GpuWrite) extends GpuV2ExistingTableWriteExec {
+
+  override def supportsColumnar: Boolean = false
+
+  override def query: SparkPlan = inner
+
+  override lazy val writingTask: GpuWritingSparkTask[_] =
```
Comment (res-life) [inline sql-plugin/src/main/spark420/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala]: [MUST FIX] Spark 4.2 still uses a metadata-aware writing task when `metadataProjection` is present: `UPDATE_OPERATION` and `COPY_OPERATION` call `write(metadata, data)`, while `INSERT_OPERATION` calls `write(data)`. Unconditionally selecting the plain task drops metadata and can replace inherited Iceberg row-lineage values with nulls.

Suggested change:
```scala
override lazy val writingTask: GpuWritingSparkTask[_] = projections match {
  case ReplaceDataProjections(dataProj, Some(metadataProj)) =>
    GpuDataAndMetadataWritingSpark420Task(dataProj, metadataProj)
  case _ =>
    GpuReplaceDataWritingSparkTask(projections)
}

// In GpuDataAndMetadataWritingSpark420Task, process contiguous operation runs in order:
// UPDATE_OPERATION | COPY_OPERATION => writer.write(metadataBatch, dataBatch)
// INSERT_OPERATION                 => writer.write(dataBatch)
// Give each run its own withRetryNoSplit scope.
```

File: integration_tests/src/main/python/iceberg/iceberg_row_lineage_test.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+ROW_LINEAGE_COLS = "_row_id, _last_updated_sequence_number"
+
+
+def _iceberg_has_row_lineage_metadata_columns():
+    """True when the Iceberg runtime exposes row-lineage metadata columns (1.9+)."""
+    try:
+        meta = spark_jvm().org.apache.iceberg.MetadataColumns
+        return meta.ROW_ID is not None and meta.LAST_UPDATED_SEQUENCE_NUMBER is not None
+    except Exception:
```
Comment (res-life) [inline integration_tests/src/main/python/iceberg/iceberg_row_lineage_test.py]: [SHOULD FIX] Catching every `Exception` turns unrelated JVM gateway or test-harness failures into an unsupported-version skip, hiding real regressions. Reflection can test API presence without suppressing unrelated failures.

Suggested change:
```diff
-    try:
-        meta = spark_jvm().org.apache.iceberg.MetadataColumns
-        return meta.ROW_ID is not None and meta.LAST_UPDATED_SEQUENCE_NUMBER is not None
-    except Exception:
-        return False
+    meta_class = spark_jvm().java.lang.Class.forName(
+        "org.apache.iceberg.MetadataColumns")
+    field_names = {field.getName() for field in meta_class.getFields()}
+    return {"ROW_ID", "LAST_UPDATED_SEQUENCE_NUMBER"}.issubset(field_names)
```

File: tests/src/test/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuDsv2MetadataWriteSuite.scala [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    val rowSchema = StructType(Seq(StructField("id", IntegerType), StructField("v", LongType)))
+    val metaSchema = StructType(Seq(StructField("_row_id", LongType)))
+    val rowProj = projectingRow(rowSchema, Seq(1, 2))
+    val metaProj = projectingRow(metaSchema, Seq(3))
+    val projections = ReplaceDataProjections(rowProj, Some(metaProj))
+
+    // writingTask selection mirrors CPU ReplaceDataExec / DataAndMetadataWritingSparkTask
+    val task = projections match {
+      case ReplaceDataProjections(dataProj, Some(metadataProj)) =>
+        GpuDataAndMetadataWritingSparkTask(dataProj, metadataProj)
```
Comment (res-life) [inline tests/src/test/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuDsv2MetadataWriteSuite.scala:81]: [SHOULD FIX] This only verifies task construction, so it does not protect the retry-boundary behavior introduced here. Add a deterministic one-shot retry test at the phase boundary.

Suggested change:
```scala
test("ordinary-row OOM does not replay a completed metadata write") {
  val writer = new RecordingDataWriter
  val task = GpuDataAndMetadataWritingSparkTask(rowProj, metaProj)
  // Build one WRITE_WITH_METADATA_OPERATION row and one WRITE_OPERATION row.
  // Inject one GpuRetryOOM when ordinary-row processing first starts.
  runTaskWrite(task, writer, mixedOperationBatch, failOrdinaryPhaseOnce = true)
  assert(writer.metadataWriteCount === 1)
  assert(writer.ordinaryWriteCount === 1)
}
```
Use the repository's existing one-shot `GpuRetryOOM` injection pattern so the failure occurs after the metadata callback succeeds.

File: integration_tests/src/main/python/iceberg/iceberg_row_lineage_test.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+    assert_equal_with_local_sort(
+        _collect_with_lineage(cpu_table),
+        _collect_with_lineage(gpu_table))
+
+
+@allow_non_gpu(*_ROW_LINEAGE_ALLOW_NON_GPU)
+@iceberg
+@ignore_order(local=True)
+def test_iceberg_row_lineage_update_cow(spark_tmp_table_factory):
```
Comment (res-life) [inline integration_tests/src/main/python/iceberg/iceberg_row_lineage_test.py]: [SHOULD FIX] This regression currently creates only unpartitioned tables, leaving the changed partitioned/clustered Iceberg lineage path uncovered.

Suggested change:
```diff
-def _create_v3_table(table_name, table_properties, df_gen=None):
+def _create_v3_table(table_name, table_properties, partition_col_sql=None, df_gen=None):
@@
-        create_iceberg_table(table_name, table_prop=props, df_gen=df_gen)
+        create_iceberg_table(table_name, partition_col_sql=partition_col_sql,
+                             table_prop=props, df_gen=df_gen)
@@
-def do_row_lineage_update_test(spark_tmp_table_factory):
+def do_row_lineage_update_test(spark_tmp_table_factory, partition_col_sql):
@@
-    _create_v3_table(cpu_table, {}, df_gen=_empty_df_gen)
-    _create_v3_table(gpu_table, {}, df_gen=_empty_df_gen)
+    _create_v3_table(cpu_table, {}, partition_col_sql, df_gen=_empty_df_gen)
+    _create_v3_table(gpu_table, {}, partition_col_sql, df_gen=_empty_df_gen)
@@
+@pytest.mark.parametrize("partition_col_sql", [None, "id"])
-def test_iceberg_row_lineage_update_cow(spark_tmp_table_factory):
+def test_iceberg_row_lineage_update_cow(spark_tmp_table_factory, partition_col_sql):
@@
-    do_row_lineage_update_test(spark_tmp_table_factory)
+    do_row_lineage_update_test(spark_tmp_table_factory, partition_col_sql)
```

File: tests/src/test/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuDsv2MetadataWriteSuite.scala [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    val seen = GpuDsv2WriteMetadata.withMetadataSchema(schema) {
+      GpuDsv2WriteMetadata.metadataSchema
+    }
+    assert(seen.contains(schema))
+    assert(GpuDsv2WriteMetadata.metadataSchema.isEmpty)
+  }
+
+  test("DeltaInsertFilter exposes distinct insert and reinsert writers") {
+    // Compile-time / API presence check: insert and reinsert helpers exist for Spark 4.x
+    assert(DeltaInsertFilter.getClass.getMethods.exists(_.getName == "writeInserts"))
```
Comment (res-life) [inline tests/src/test/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuDsv2MetadataWriteSuite.scala]: [SHOULD FIX] Reflection only proves these helper names exist; it still passes if reinserts use the wrong filter, call `insert`, or drop metadata. Replace it with behavioral coverage.

Suggested change:
```scala
test("DeltaInsertFilter routes insert and reinsert rows with metadata") {
  val writer = new RecordingDeltaWriter
  withResource(buildOperationBatch(INSERT_OPERATION, REINSERT_OPERATION, INSERT_OPERATION)) { b =>
    DeltaInsertFilter.writeInserts(writer, b, rowProjection, rowDataTypes)
    DeltaInsertFilter.writeReinserts(
      writer, b, rowProjection, rowDataTypes, metadataProjection, metadataDataTypes)
  }
  assert(writer.insertedRows === 2)
  assert(writer.reinsertedRows === 1)
  assert(writer.reinsertMetadataRows === 1)
  assert(writer.reinsertSchema.contains(metadataProjection.schema))
}
```
The recording writer should close accepted batches and capture `GpuDsv2WriteMetadata.metadataSchema` inside `reinsert`.

File: integration_tests/src/main/python/iceberg/iceberg_row_lineage_test.py
Comment (liurenjie1024) [inline integration_tests/src/main/python/iceberg/iceberg_row_lineage_test.py:1]: I would prefer to drop support of this iceberg related changes. It's part of iceberg v3, and we are currently not   supporting it yet.

File: sql-plugin/src/main/spark350/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala (hunk 3 above)
Comment (firestarman) [reply sql-plugin/src/main/spark350/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: Done: split Delta writeInserts/writeReinserts into separate withRetryNoSplit scopes so a later-phase OOM retries only that phase.

File: sql-plugin/src/main/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala (hunk 4 above)
Comment (firestarman) [reply sql-plugin/src/main/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala]: Done: process ReplaceData via splitIntoContiguousOperationRuns so metadata/ordinary writes stay in input operation order.

File: sql-plugin/src/main/spark420/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala (hunk 5 above)
Comment (firestarman) [reply sql-plugin/src/main/spark420/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala]: Done: Spark 4.2 now uses GpuDataAndMetadataWritingSpark420Task for UPDATE/COPY/INSERT metadata routing when metadataProjection is present.

File: integration_tests/src/main/python/iceberg/iceberg_row_lineage_test.py (hunk 6 above)
Comment (firestarman) [reply integration_tests/src/main/python/iceberg/iceberg_row_lineage_test.py]: Done: switched to Class.forName reflection for API presence instead of catching broad Exception as unsupported-version skip.

File: tests/src/test/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuDsv2MetadataWriteSuite.scala (hunk 7 above)
Comment (firestarman) [reply tests/src/test/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuDsv2MetadataWriteSuite.scala:81]: Done: added a deterministic one-shot OOM/retry unit test at the phase boundary in GpuDsv2MetadataWriteSuite.

File: integration_tests/src/main/python/iceberg/iceberg_row_lineage_test.py (hunk 8 above)
Comment (firestarman) [reply integration_tests/src/main/python/iceberg/iceberg_row_lineage_test.py]: Done: partitioned IT now parametrizes partition_cols as None/"id" to cover unpartitioned and partitioned Iceberg lineage paths.

File: tests/src/test/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuDsv2MetadataWriteSuite.scala (hunk 9 above)
Comment (firestarman) [reply tests/src/test/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuDsv2MetadataWriteSuite.scala]: Done: replaced name-only reflection with a DeltaInsertFilter behavioral unit test covering reinsert filter/metadata behavior.

File: integration_tests/src/main/python/iceberg/iceberg_row_lineage_test.py
Comment (firestarman) [reply integration_tests/src/main/python/iceberg/iceberg_row_lineage_test.py:1]: Done. Dropped the Iceberg V3 changes; the PR now keeps only generic DSv2 metadata/reinsert support.

File: sql-plugin/src/main/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala [hunk 10]
```diff
@@ -73,3 +85,63 @@ case class GpuReplaceDataWritingSparkTask(
...
+  extends GpuWritingSparkTask[DataWriter[ColumnarBatch]] {
+
+  private lazy val rowProjection = GpuProjectingColumnarBatch(dataProj)
+  private lazy val metadataProjection = GpuProjectingColumnarBatch(metadataProj)
+
+  override protected def write(
+      writer: DataWriter[ColumnarBatch],
+      batch: ColumnarBatch): Unit = {
+    // Split only where column 0 changes; do not group equal operations globally.
+    withResource(splitIntoContiguousOperationRuns(batch)) { runs =>
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/spark400/scala/org/apache/spark/sql/execution/datasources/v2/GpuReplaceDataExec.scala:107]: Fixed in d0afe0f6d. `splitIntoContiguousOperationRuns` now makes the raw input spillable before entering `withRetryNoSplit`, so boundary discovery and `contiguousSplit` run inside an idempotent retry scope before any writer side effects. I also added a GPU OOM injection test that verifies the split retries and preserves the metadata/ordinary write order without duplicate writes. Spark 4.0/4.2 scalastyle validation and the focused six-test suite pass.

**Comment by firestarman** [comment]
build

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM. Iceberg is currently affected by this component, please run iceberg it locally to verify that it works.

**Comment by firestarman** [comment]
I am going to close this, for details please go to the linked issue.
